cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/StatefulSetTwillPreparer.java

```diff
@@ -68,46 +69,51 @@
   private final PodInfo podInfo;
   private final Map<String, String> cConf;
 
-  StatefulSetTwillPreparer(ApiClient apiClient, String kubeNamespace, PodInfo podInfo, TwillSpecification spec,
-                           RunId twillRunId, V1ObjectMeta resourceMeta, Map<String, String> cConf,
-                           KubeTwillControllerFactory controllerFactory) {
-    super(apiClient, kubeNamespace, podInfo, spec, twillRunId, resourceMeta, cConf, controllerFactory);
+  StatefulSetTwillPreparer(Map<String, String> cConf, ApiClient apiClient, String kubeNamespace, PodInfo podInfo,
+                           TwillSpecification spec, RunId twillRunId, String resourcePrefix,
+                           Map<String, String> extraLabels, KubeTwillControllerFactory controllerFactory) {
+    super(cConf, apiClient, kubeNamespace, podInfo, spec, twillRunId, resourcePrefix, extraLabels, V1StatefulSet.class,
+          controllerFactory);
     this.podInfo = podInfo;
     this.cConf = cConf;
   }
 
   @Override
-  protected void createKubeResources(V1ObjectMeta resourceMeta, V1ResourceRequirements resourceRequirements,
-                                     List<V1EnvVar> envVars, long timeout,
-                                     TimeUnit timeoutUnit) throws IOException, ApiException {
+  protected V1ObjectMeta createKubeResources(V1ObjectMeta resourceMeta, V1ResourceRequirements resourceRequirements,
+                                             List<V1EnvVar> envVars, long timeout, TimeUnit timeoutUnit)
+    throws IOException, ApiException {
     AppsV1Api appsApi = new AppsV1Api(getApiClient());
     V1StatefulSet statefulSet = createStatefulSet(resourceMeta, resourceRequirements, envVars,
                                                   timeoutUnit.toMillis(timeout));
     LOG.info("Creating StatefulSet with spec to K8s {}", getApiClient().getJSON().serialize(statefulSet));
     try {
-      appsApi.createNamespacedStatefulSet(getKubeNamespace(), statefulSet, "true", null, null);
+      statefulSet = appsApi.createNamespacedStatefulSet(getKubeNamespace(), statefulSet, "true", null, null);
     } catch (ApiException e) {
       if (e.getCode() != HttpURLConnection.HTTP_CONFLICT) {
         throw e;
       }
     }
     LOG.info("Created StatefulSet with spec to K8s {}", getApiClient().getJSON().serialize(statefulSet));
+    return statefulSet.getMetadata();
   }
 
   private V1StatefulSet createStatefulSet(V1ObjectMeta resourceMeta, V1ResourceRequirements resourceRequirements,
                                           List<V1EnvVar> envVars, long startTimeoutMillis) {
+    // Update owner reference for the statefulset to be current pod
+    // so that pod deletion triggers deletion of statefulsets too.
+    V1OwnerReference ownerReference = new V1OwnerReference()
+      .apiVersion(podInfo.getApiVersion())
+      .kind(podInfo.getKind())
+      .name(podInfo.getName())
+      .uid(podInfo.getUid())
+      .blockOwnerDeletion(true);
```

Comment:
Setting up the owner reference for the preview runner statefulset as the preview manager pod so that when that pod gets deleted we recreate stateful sets.

Reply:
Why the owner is the pod?? If the pod deleted then the runner statefulset will be deleted?? If that what you wanted? If the pod is deleted manually to force a restart, the preview runners statefulset will be deleted too?

Reply:
Yes this is done intentionally. Preview runner statefulset does not store any state anyway. Deleting the runners when preview manager deleted make sure both are running with same version of the code. Also in absence of this deletion, preview manager can create new statefulset for runner and pods from which can try polling for preview request for short amount of time.

Reply:
So, this means if the preview manager pod get restarted due to failure node or OOM, then it kills all running previews? Seems like not a very good design.

Reply:
As discussed offline filed an improvement here - https://issues.cask.co/browse/CDAP-17188